services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -0,0 +1,180 @@
+
+    /**
+     * The subscription received from the publisher this subscriber subscribes to.
+     */
+    private Subscription subscription;
+
+    /**
+     * This future will be completed once this subscriber reaches a terminal state, failed or successfully, and will be completed
+     * accordingly.
+     */
+    private final CompletableFuture<Void> future = new CompletableFuture<>();
+
+    private final Object lock = new Object();
+
+    /**
+     * The etag of the object being downloaded.
+     */
+    private String eTag;
+
+    public MultipartDownloaderSubscriber(S3AsyncClient s3, GetObjectRequest getObjectRequest) {
+        this(s3, getObjectRequest, 0);
+    }
+
+    public MultipartDownloaderSubscriber(S3AsyncClient s3, GetObjectRequest getObjectRequest, int completedParts) {
+        this.s3 = s3;
+        this.getObjectRequest = getObjectRequest;
+        this.completedParts = new AtomicInteger(completedParts);
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        if (this.subscription != null) {
+            s.cancel();
+            return;
+        }
+        this.subscription = s;
+        this.subscription.request(1);
+    }
+
+    @Override
+    public void onNext(AsyncResponseTransformer<GetObjectResponse, GetObjectResponse> asyncResponseTransformer) {
+        log.trace(() -> "onNext " + completedParts.get());
+        if (asyncResponseTransformer == null) {
+            throw new NullPointerException("onNext must not be called with null asyncResponseTransformer");
+        }
+
+        int nextPartToGet = completedParts.get() + 1;
+
+        if (totalParts != null && nextPartToGet > totalParts) {
+            synchronized (lock) {
+                subscription.cancel();
+            }
+        }
+
+        GetObjectRequest actualRequest = nextRequest(nextPartToGet);
+        CompletableFuture<GetObjectResponse> getObjectFuture = s3.getObject(actualRequest, asyncResponseTransformer);
+        getObjectFuture.whenComplete((response, error) -> {
+            if (error != null) {
+                onError(error);
+                return;
+            }
+            requestMoreIfNeeded(response);
+        });
+    }
+
+    private void requestMoreIfNeeded(GetObjectResponse response) {
+        int totalComplete = completedParts.incrementAndGet();
+        MultipartDownloadUtils.multipartDownloadResumeContext(getObjectRequest)
+                              .ifPresent(ctx -> {
+                                  ctx.addCompletedPart(totalComplete);
+                                  ctx.addToBytesToLastCompletedParts(response.contentLength());
+                                  if (ctx.response() == null) {
+                                      ctx.response(response);
+                                  }
+                              });
+        log.trace(() -> String.format("completed part: %d", totalComplete));
+
+        if (eTag == null) {
+            this.eTag = response.eTag();
```

Comment:
Is this etag for the entire object or a part? Also, it needs to be volatile

Reply:
entire object, adding volatile